### PR TITLE
Merge ONNX decoder models

### LINKED ISSUE #530: Merge decoder and decoder with past to reduce memory usage
### Feature request

* A function to merge decoder and decoder with past after their export.
* ORTModel for CausalML and Seq2SeqML need refactoring to take only the merged decoder.

### Motivation

__Context__

Currently while using ONNX models for the decoding process, Optimum uses two decoder models: the decoder and decoder with past. This means that redundant initializers are loaded twice which is not optimal in terms of memory usage.

### Your contribution

I will open a PR for it.

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?

- [x] Merge `decoder` and `decoder_with_past` in a single model proto
- [x] Check what opset to use
- [x] Apply ONNXSIM to simplify subgraphs
- [x] Cross-subgraphs initializers deduplication for the merged model

Next steps:

- [ ] Need to check if it works for large proto, and how shall we export it(but the support will be added in another PR)
- [ ] Need to adapt ORTModels to be able to take benefits(New input `use_cache`, dummy inputs for `past_key_values`...)
- [ ] Need to check if merged ONNX model works for IOBinding (introduces new input `use_cache`, but dlpack doesn't support `dtype=bool` ) 
- [ ] Open discussion: Should it be applied automatically after export or before inference? (Guess no, optimizer won't be able to due with cross-subgraphs optimization)

Fixes #530 

--------------------------------------

Snippets to test the merge:

```python
# Load decoders
decoder = onnx.load(decoder_path)
decoder_with_past = onnx.load(decoder_with_past_path)

merged_model = merge_decoders(decoder, decoder_with_past)
onnx.save(merged_model, merged_model_path)
```

And test inference:
```python
from transformers import AutoConfig, AutoTokenizer
from optimum.onnxruntime import ORTModelForCustomTasks

model_id = "gpt2"
tokenizer = AutoTokenizer.from_pretrained(model_id)
payload = tokenizer("I love burritos!", return_tensors="pt")
onnx_inputs = {
    "input_ids": payload.input_ids.cpu().detach().numpy(),
    "attention_mask": payload.attention_mask.cpu().detach().numpy(),
    "use_cache": np.full(1, True),
}
shape = (1, 12, 0, 64)
for i in range(24):
    onnx_inputs[f"past_key_values_{i}"] = np.random.uniform(low=0.5, high=13.3, size=shape).astype(np.float32)

sess = ort.InferenceSession(merged_model_path, providers=["CPUExecutionProvider"])
pred = sess.run(None, onnx_inputs)
```

--------------------------------------
After the merge, the merged decoder shall have almost the same size as decoder or decoder with past

* GPT2
image

* T5 small

image

### CODE REVIEW

**Comment by fxmarty** [comment]
Will likely help for https://discuss.huggingface.co/t/reducing-latency-for-gpt-j/27547

File: optimum/onnx/graph_transformations.py [hunk 1]
```diff
@@ -79,3 +91,217 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
 
     return model
+
+
+def replace_atenops_to_gather(model: ModelProto):
+    """
+    Replaces broken ATenOp nodes back to Gather nodes.
+
+    Args:
+        model (`~onnx.ModelProto`): The ONNX model to fix.
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:106]:
```suggestion
        model (`onnx.ModelProto`): 
            The ONNX model to fix.
```

File: optimum/onnx/graph_transformations.py [hunk 2]
```diff
@@ -79,3 +91,217 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+        inplace (`bool`, defaults to False): Whether to perform this transformation inplace.
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:107]:
```suggestion
        inplace (`bool`, defaults to `False`): 
            Whether to perform this transformation inplace.
```

File: optimum/onnx/graph_transformations.py [hunk 3]
```diff
@@ -79,3 +91,217 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+        else:
+            raise ValueError(f"Can not find `dim_param` nor `dim_value` in the output dimension info.")
+
+    return output_shape
+
+
+def _check_num_outputs(model1: ModelProto, model2: ModelProto):
+    if not len(model1.graph.output) == len(model2.graph.output):
+        raise ValueError(
+            f"Two model protos need to have same outputs. But one has {len(model1.graph.output)} "
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:147]:
```suggestion
            f"Two model protos need to have the same outputs. But one has {len(model1.graph.output)} "
```

File: optimum/onnx/graph_transformations.py [hunk 4]
```diff
@@ -79,3 +91,217 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+            f"outputs while the other has {len(model2.graph.output)} outputs."
+        )
+
+
+def _unify_onnx_outputs(model1: ModelProto, model2: ModelProto):
+    """
+    Unify the outputs of deux ONNX model protos. The outputs of model1 will be replaced by outputs of model2.
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:154]:
```suggestion
    Unifies the outputs of deux ONNX model protos. The outputs of model1 will be replaced by outputs of model2.
```

File: optimum/onnx/graph_transformations.py [hunk 5]
```diff
@@ -79,3 +91,217 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+    if not all(
+        model_output_1 == model_output_2
+        for model_output_1, model_output_2 in zip(model1.graph.output, model2.graph.output)
+    ):
+        raise RuntimeError(f"Failed to unify outputs of given ONNX model protos.")
+
+
+def _get_all_inputs(model_list: List[ModelProto]):
+    inputs = []
+    input_names = []
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:188]:
```suggestion
    input_names = set()
```

File: optimum/onnx/graph_transformations.py [hunk 6]
```diff
@@ -79,3 +91,217 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+    for model in model_list:
+        for input in model.graph.input:
+            if input.name not in input_names:
+                input_names.append(input.name)
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:192]:
```suggestion
                input_names.add(input.name)
```

File: optimum/onnx/graph_transformations.py [hunk 7]
```diff
@@ -79,3 +91,217 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+
+def _get_onnx_opset(model: ModelProto):
+    opset_import = model.opset_import[0]
+    return getattr(opset_import, "version")
+
+
+def simplify_onnx_graph(model: ModelProto, print_info: bool = True) -> ModelProto:
+
+    model_simp, check = simplify(model)
+    assert check, "Simplified ONNX model could not be validated."
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:205]: In general it is better to raise a proper exception than using asser (except in tests) IMO

File: optimum/onnx/graph_transformations.py [hunk 8]
```diff
@@ -79,3 +91,217 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+    if print_info:
+        print_simplifying_info(model, model_simp)
+
+    return model_simp
+
+
+def _deduplicated_cross_model_initializers(models: List[ModelProto], suffix: str = None):
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:212]:
```suggestion
def _deduplicated_cross_model_initializers(models: List[ModelProto], suffix: str = ""):
```

File: optimum/onnx/graph_transformations.py [hunk 9]
```diff
@@ -79,3 +91,217 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+
+    all_initializers = list()
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:214]:
```suggestion
    all_initializers = []
```

File: optimum/onnx/graph_transformations.py [hunk 10]
```diff
@@ -79,3 +91,217 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+    for model in models:
+        all_initializers += list(model.graph.initializer)
+
+    duplicates = _find_duplicate_initializers(all_initializers)
+    name_sharing_dict = _create_name_sharing_dict(duplicates, suffix=suffix)
+    for model in models:
+        _replace_input_names(model, name_sharing_dict)
+
+    deduplicated_initializers = list()
+    deduplicated_name = list()
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:224]:
```suggestion
    deduplicated_initializers = []
    deduplicated_name = []
```

File: optimum/onnx/graph_transformations.py [hunk 11]
```diff
@@ -79,3 +91,217 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+
+    Args:
+        decoder (`~onnx.ModelProto`): Decoder ONNX model.
+        decoder_with_past (`~onnx.ModelProto`): Decoder with past ONNX model.
+        graph_name (`~str`): Name of the parent graph(graph of the control flow node).
+        producer_name (`~str`): Graph producer name.
+        simplify_graph (`~bool`): Whether to use onnx-simplifier to simplify subgraphs.
+
+    Returns:
+        `~onnx.ModelProto`: The fused decoder ONNX model.
```
Comment (michaelbenayoun) [inline optimum/onnx/graph_transformations.py:253]: Same comment

**Review by michaelbenayoun — COMMENTED** [review]
Left mostly nits!

File: optimum/onnx/graph_transformations.py [hunk 12]
```diff
@@ -30,11 +37,16 @@ def _find_duplicate_weights(model) -> DefaultDict[Tuple[int, bytes], Set[str]]:
     return duplicates
 
 
-def _create_name_sharing_dict(duplicate_weights: DefaultDict[Tuple[int, bytes], Set[str]]) -> Dict[str, str]:
+def _create_name_sharing_dict(
+    duplicate_weights: DefaultDict[Tuple[int, bytes], Set[str]], suffix: str = None
+) -> Dict[str, str]:
     def _create_name_sharing_dict_for_duplicates(duplicates: Set[str]) -> Dict[str, str]:
         common_name = duplicates.pop()
         duplicates.add(common_name)
```
[the remark this answers is not in the document]
Comment (JingyaHuang) [reply optimum/onnx/graph_transformations.py:50]: Nope, it's good. I just want to reuse this function and I need to add a suffix option to avoid decoder / decoder with past naming collision.

File: optimum/onnx/graph_transformations.py (hunk 4 above)
Comment (JingyaHuang) [reply optimum/onnx/graph_transformations.py:156]: You didn't find that I wrote "two" in French :<

**Comment by fxmarty** [comment]
```python

from optimum.onnx.graph_transformations import merge_decoders

import onnx

decoder_path = "/home/fxmarty/hf_internship/optimum/t5_onnx/decoder_model.onnx"
decoder_with_past_path = "/home/fxmarty/hf_internship/optimum/t5_onnx/decoder_with_past_model.onnx"

# Load decoders
decoder = onnx.load(decoder_path)
decoder_with_past = onnx.load(decoder_with_past_path)

merged_model = merge_decoders(decoder, decoder_with_past)
onnx.save(merged_model, merged_model_path)
```

raises:

```
Traceback (most recent call last):
  File "<tmp 5>", line 43, in <module>
    merged_model = merge_decoders(decoder, decoder_with_past)
  File "/home/fxmarty/hf_internship/optimum/optimum/onnx/graph_transformations.py", line 260, in merge_decoders
    _unify_onnx_outputs(decoder, decoder_with_past)
  File "/home/fxmarty/hf_internship/optimum/optimum/onnx/graph_transformations.py", line 167, in _unify_onnx_outputs
    raise ValueError(
ValueError: Can not match 1233 with 1174. Make sure your model protos have same outputs, have same data types and are in the same order.
```

Transformers commit: 3ee958207a9830da6afcf2c20772ef9a2159ca30
Optimum commit: 9c71dd52f71f397bd97da2228691a3c7049f4452

Model exported with:

```
python3 -m optimum.exporters.onnx --model t5-small --task seq2seq-lm-with-past --for-ort t5_onnx
```

**Comment by fxmarty** [comment]
Can you add merge_decoders in `onnx/__init__.py`?

File: optimum/onnx/graph_transformations.py [hunk 13]
```diff
@@ -79,3 +96,219 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+    if not len(model1.graph.output) == len(model2.graph.output):
+        raise ValueError(
+            f"Two model protos need to have the same outputs. But one has {len(model1.graph.output)} "
+            f"outputs while the other has {len(model2.graph.output)} outputs."
+        )
+
+
+def _unify_onnx_outputs(model1: ModelProto, model2: ModelProto):
+    """
+    Unifies the outputs of deux ONNX model protos. The outputs of model1 will be replaced by outputs of model2.
```
Comment (fxmarty) [inline optimum/onnx/graph_transformations.py]:
```suggestion
    Unifies the outputs of two ONNX model protos. The outputs of model1 will be replaced by outputs of model2.
```

File: optimum/onnx/graph_transformations.py [hunk 14]
```diff
@@ -79,3 +96,219 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+        graph_name (`~str`): Name of the parent graph(graph of the control flow node).
+        producer_name (`~str`): Graph producer name.
+        simplify_graph (`~bool`): Whether to use onnx-simplifier to simplify subgraphs.
+
+    Returns:
+        `~onnx.ModelProto`: The fused decoder ONNX model.
+    """
+    if simplify_graph:
+        decoder = simplify_onnx_graph(decoder)
+        decoder_with_past = simplify_onnx_graph(decoder_with_past)
```
Comment (fxmarty) [inline optimum/onnx/graph_transformations.py:259]: Is this related to merging decoder/decoder with past? If yes, why is it needed?

File: optimum/onnx/graph_transformations.py [hunk 15]
```diff
@@ -79,3 +96,219 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+        outputs=no_past_branch.output,
+        initializer=deduplicated_initializers,
+    )
+    merged_model = onnx.helper.make_model(
+        merged_graph,
+        producer_name=producer_name,
+        opset_imports=[
+            onnx.helper.make_opsetid(
+                domain=onnx.defs.ONNX_DOMAIN,
+                version=_get_onnx_opset(decoder),
```
Comment (fxmarty) [inline optimum/onnx/graph_transformations.py:308]: what happens if decoder and decoder_with_past use a different opset? Is there a check for it?

File: optimum/onnx/graph_transformations.py [hunk 16]
```diff
@@ -79,3 +96,219 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+        simplify_graph (`~bool`): Whether to use onnx-simplifier to simplify subgraphs.
+
+    Returns:
+        `~onnx.ModelProto`: The fused decoder ONNX model.
+    """
+    if simplify_graph:
+        decoder = simplify_onnx_graph(decoder)
+        decoder_with_past = simplify_onnx_graph(decoder_with_past)
+
+    _unify_onnx_outputs(decoder, decoder_with_past)
```
Comment (fxmarty) [inline optimum/onnx/graph_transformations.py:261]: just to understand, why do we do this? Is it so that the decoder spits out the past key values? I think a comment would help

File: optimum/onnx/graph_transformations.py (hunk 8 above)
Comment (JingyaHuang) [reply optimum/onnx/graph_transformations.py:215]: I have checked if the suffix is None or not in the function `_create_name_sharing_dict`, if we give a "", it will always add a "_" after the original name and it will be ugly.

File: optimum/onnx/graph_transformations.py (hunk 14 above)
Comment (JingyaHuang) [reply optimum/onnx/graph_transformations.py:261]: It will fail during the checker or during inference. Good point, I will add a check here.

File: optimum/onnx/graph_transformations.py (hunk 16 above)
Comment (JingyaHuang) [reply optimum/onnx/graph_transformations.py:263]: According to the rules of "If" op, two subgraphs must have the same number of outputs. Comment added to this function.

File: optimum/onnx/graph_transformations.py (hunk 15 above)
Comment (JingyaHuang) [reply optimum/onnx/graph_transformations.py:310]: It will fail during the checker or during inference. Good point, I will add a check here.

File: optimum/onnx/graph_transformations.py (hunk 14 above)
Comment (JingyaHuang) [reply optimum/onnx/graph_transformations.py:261]: onnx-optimizer and onnx-simplifier are tools recommended by ONNX, which reduce ops in conservative ways with onnxruntime. I find it easier to use than ORTOptimizer.

File: optimum/onnx/graph_transformations.py (hunk 15 above)
Comment (JingyaHuang) [reply optimum/onnx/graph_transformations.py:310]: But not sure if we want to introduce another tool that overlaps a bit with what we have.

File: optimum/onnx/graph_transformations.py (hunk 14 above)
Comment (JingyaHuang) [reply optimum/onnx/graph_transformations.py:261]: But not sure if we want to introduce another tool that overlaps a bit with what we have.

File: optimum/onnx/graph_transformations.py (hunk 15 above)
Comment (JingyaHuang) [reply optimum/onnx/graph_transformations.py:310]: Added!

**Comment by JingyaHuang** [comment]
> python3 -m optimum.exporters.onnx --model t5-small --task seq2seq-lm-with-past --for-ort t

> ```python
> from optimum.onnx.graph_transformations import merge_decoders
> 
> import onnx
> 
> decoder_path = "/home/fxmarty/hf_internship/optimum/t5_onnx/decoder_model.onnx"
> decoder_with_past_path = "/home/fxmarty/hf_internship/optimum/t5_onnx/decoder_with_past_model.onnx"
> 
> # Load decoders
> decoder = onnx.load(decoder_path)
> decoder_with_past = onnx.load(decoder_with_past_path)
> 
> merged_model = merge_decoders(decoder, decoder_with_past)
> onnx.save(merged_model, merged_model_path)
> ```
> 
> raises:
> 
> ```
> Traceback (most recent call last):
>   File "<tmp 5>", line 43, in <module>
>     merged_model = merge_decoders(decoder, decoder_with_past)
>   File "/home/fxmarty/hf_internship/optimum/optimum/onnx/graph_transformations.py", line 260, in merge_decoders
>     _unify_onnx_outputs(decoder, decoder_with_past)
>   File "/home/fxmarty/hf_internship/optimum/optimum/onnx/graph_transformations.py", line 167, in _unify_onnx_outputs
>     raise ValueError(
> ValueError: Can not match 1233 with 1174. Make sure your model protos have same outputs, have same data types and are in the same order.
> ```
> 
> Transformers commit: 3ee958207a9830da6afcf2c20772ef9a2159ca30 Optimum commit: [9c71dd5](https://github.com/huggingface/optimum/commit/9c71dd52f71f397bd97da2228691a3c7049f4452)
> 
> Model exported with:
> 
> ```
> python3 -m optimum.exporters.onnx --model t5-small --task seq2seq-lm-with-past --for-ort t5_onnx
> ```

Due to previous export bug. All good now.

File: optimum/onnx/graph_transformations.py (hunk 8 above)
Comment (michaelbenayoun) [reply optimum/onnx/graph_transformations.py:217]: Alright!

**Review by michaelbenayoun — APPROVED** [review]
LGTM!

File: tests/onnx/test_onnx_graph_transformations.py [hunk 17]
```diff
@@ -58,5 +62,24 @@ def test_weight_sharing_output_match(self):
             )
 
 
+class OnnxMergingTestCase(TestCase):
+    SUPPORTED_ARCHITECTURES_WITH_MODEL_ID = {
+        "gpt2": "optimum/gpt2",
```
Comment (fxmarty) [inline tests/onnx/test_onnx_graph_transformations.py:67]: I would use tiny models here, and use models exported using `optimum.exporters.onnx` (for example this test did not catch the problem in https://github.com/huggingface/optimum/pull/601 )

File: optimum/onnx/graph_transformations.py (hunk 14 above)
Comment (fxmarty) [reply optimum/onnx/graph_transformations.py:262]: What was the motive to introduce it in this PR? I think it's fine to keep, just that I don't see the link with the PR

File: optimum/onnx/graph_transformations.py [hunk 18]
```diff
@@ -79,3 +97,227 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+        decoder (`onnx.ModelProto`): Decoder ONNX model.
+        decoder_with_past (`onnx.ModelProto`): Decoder with past ONNX model.
+        graph_name (`str`): Name of the parent graph(graph of the control flow node).
+        producer_name (`str`): Graph producer name.
+        simplify_graph (`bool`): Whether to use onnx-simplifier to simplify subgraphs.
+
+    Returns:
+        `~onnx.ModelProto`: The fused decoder ONNX model.
+    """
+    if simplify_graph:
```
Comment (fxmarty) [inline optimum/onnx/graph_transformations.py:260]: can we raise an error if `not is_onnxsim_available()`?

File: tests/onnx/test_onnx_graph_transformations.py (hunk 17 above)
Comment (JingyaHuang) [reply tests/onnx/test_onnx_graph_transformations.py:67]: Do you mean that we should have assets for tiny models exported with `optimum.exporters.onnx`?

Comment (fxmarty) [reply tests/onnx/test_onnx_graph_transformations.py:67]: No, I mean using models exported on the fly during the test `optimum.exporters.onnx`.

**Comment by JingyaHuang** [comment]
The test for t5 can pass only of #601 is merged.

**Comment by fxmarty** [comment]
Merged!

File: optimum/onnx/graph_transformations.py [hunk 19]
```diff
@@ -79,3 +90,211 @@ def remove_duplicate_weights(model: ModelProto, inplace: bool = False) -> ModelP
...
+
+def replace_atenops_to_gather(model: ModelProto):
+    """
+    Replaces broken ATenOp nodes back to Gather nodes.
+
+    Args:
+        model (`onnx.ModelProto`):
+            The ONNX model to fix.
+        inplace (`bool`, defaults to `False`):
+            Whether to perform this transformation inplace.
```
Comment (fxmarty) [inline optimum/onnx/graph_transformations.py:103]:
```suggestion
```

**Review by fxmarty — APPROVED** [review]
Awesome!

**Comment by Eichhof** [comment]
I tried it on GPT-J to merge the models and I'm getting the error `Detected 1 oom-kill event(s) in StepId=5333998.batch. Some of your processes may have been killed by the cgroup out-of-memory handler.` although was running the process with 150GB of memory.

File: optimum/onnx/graph_transformations.py [hunk 20]
```diff
@@ -13,28 +13,39 @@
...
-    for initializer in model.graph.initializer:
+    for initializer in initializers:
+        tensor_dims = tuple(getattr(initializer, "dims"))
         for data_attr in ["raw_data", "int32_data", "int64_data", "uint64_data", "float_data", "double_data"]:
             tensor_data = getattr(initializer, data_attr)
             if tensor_data:
                 tensor_data = tuple(tensor_data)
                 break
-        duplicates[(initializer.data_type, tensor_data)].add(initializer.name)
+        duplicates[(initializer.data_type, tensor_data, tensor_dims)].add(initializer.name)
```
Comment (JingyaHuang) [inline optimum/onnx/graph_transformations.py:35]: Here I addded tensor_dims, I doubt if I met some small initializers(constant) with the same tensor_data but different data_type / tensor_dims.

Comment (fxmarty) [reply optimum/onnx/graph_transformations.py:35]: Maybe it's best to keep as a tuple then until we validate further. The `to_array` + hash trick seem to work fine both memory-wise and time-wise.
